### PR TITLE
Store: If TaxJar plugin is active, do not show core settings

### LINKED ISSUE #17110: WooCommerce - controls on the tax settings page always revert
The controls on the tax settings page always revert to their previous state.

To reproduce, change one of the controls, save, and then refresh the page after a couple of seconds (revert is not always instantaneous) 
* "Tax calculations enabled" toggle will always be checked
* "Taxes are included in product prices" will always be unchecked
* "Charge taxes on shipping costs" will always be checked

We want TaxJar to always be enabled, and the plugin forces certain settings to revert to predefined values. For more info see p1502385490878365-slack-woocommerce-on-wpcom

We should consider removing these controls from the tax settings page.

CC @allendav @kellychoffman @jameskoster

### PR DESCRIPTION
**Pull request by allendav**
Fixes #17110 

To test:
* Navigate to http://calypso.localhost:3000/store/settings/taxes/{DOMAIN}
* Wait for the placeholder to pass
* You'll either be placed on a view that tells you you have the TaxJar plugin active, or a view that lets you adjust your tax settings
* Navigate to http://calypso.localhost:3000/plugins/manage/{DOMAIN} and deactivate TaxJar if active, or install and activate if not
* Return to  http://calypso.localhost:3000/store/settings/taxes/{DOMAIN}
* Ensure you see whatever view you didn't see before

cc @jeffstieler @justinshreve @kellychoffman @jameskoster 

edit: Note: I am going to submit a PR to TaxJar next to have them _not_ hook :allofthethings: unless the user has both checked the enabled setting for their plugin AND entered an API key. That should help WCS easily work around active TaxJar plugins that are not fully configured with an API key.

screen shot 2017-11-06 at 12 46 17 pm

screen shot 2017-11-06 at 12 47 02 pm

### CODE REVIEW

File: client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+				<div className="taxes__tax-jar">
+					<ExtendedHeader
+						label={ translate( 'Tax Plugin Detected' ) }
+						description={ translate( "A third-party plugin is managing your store's taxes." ) }
+					/>
+					<Card>
+						<p>
+							{ translate( "Your store's taxes are being managed by a third-party plugin. " ) }
+							<ExternalLink icon href={ adminUrl } rel="noopener noreferrer">
+								{ translate( 'Plugin settings' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js]: Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Edit plugin settings' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+		event.preventDefault();
+		this.setState( { isSaving: true } );
+
+		const onSuccess = () => {
+			this.setState( { isSaving: false, userBeganEditing: false } );
+			if ( onSuccessExtra ) {
+				onSuccessExtra();
+			}
+			return successNotice( translate( 'Settings updated successfully.' ), {
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js]: Hi! I've found a possible matching string that has already been translated **24** times:
`translate( 'Theme settings updated successfully.' )` *ES Score: 13*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/17110-store-tax-settings-detect-taxjar)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by jameskoster** [comment]
OK, I need a rundown on this cause it feels a little weird how I understand it currently.

The issue is the TaxJar plugin doesn't work properly in Calypso and has been superseded by functionality in services, right?

So, if the TaxJar plugin is active, you see this;

screen shot 2017-11-06 at 12 46 17 pm

And if it's not (services is handling taxes) you see this;

screen shot 2017-11-06 at 12 47 02 pm

Is that right? If so, the flow of heading over to manage plugins, deactivating TaxJar, then coming back to the tax settings screen to learn that my taxes are now being handled by........ TaxJar... would feel very odd xD

Just want to make sure my understanding is correct before giving feedback.

**Comment by allendav** [comment]
@jameskoster 

It is a little odd - it is because we _used_ to install the TaxJar plugin for people and then hooked it to allow WooCommerce Services to drive requests through WooCommerce Services' server to the TaxJar server.

Now, WooCommerce Services can do tax calculations without the TaxJar plugin installed (and yet it still uses TaxJar on the backend so we need to give the data privacy notice.)

We still want to allow people to continue to use the TaxJar plugin if they want to (e.g. for filing their taxes electronically) so that's basically why we have these two views (one view with TaxJar plugin active and one view without)

So... all that said, I would like us to consider replacing (for the "without" case)

> Sales tax calculations are provided by a third party: TaxJar. By enabling this option, TaxJar will have access to some of your data. Learn more.

with

> Sales tax calculations are provided by WooCommerce Services. When this option is enabled, WooCommerce Services will share some of your order information with a third party. Learn more.

cc @jeffstieler @pmaiorana @justinshreve

**Comment by justinshreve** [comment]
> Sales tax calculations are provided by WooCommerce Services. When this option is enabled, WooCommerce Services will share some of your data with a third party. Learn more.

I think that text makes sense. It's definitely less confusing.

**Comment by allendav** [comment]
@justinshreve wrote:

> > Sales tax calculations are provided by WooCommerce Services. When this option is enabled, WooCommerce Services will share some of your data with a third party. Learn more.

> I think that text makes sense. It's definitely less confusing.

Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/9d7031605143634a0dedfe8e5aeea5ec90cd5a15 looks like this now @jameskoster 

screen shot 2017-11-13 at 11 49 44 am

File: client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+				onCheckboxChange={ this.onCheckboxChange }
+				pricesIncludeTaxes={ this.state.pricesIncludeTaxes }
+				shippingIsTaxable={ this.state.shippingIsTaxable }
+			/>
+		);
+	};
+
+	render = () => {
+		const { className, loaded, site, translate } = this.props;
+
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js]: No possibility, but I just made it isRequired in https://github.com/Automattic/wp-calypso/pull/19535/commits/054bd7f282edba71c5f26c8377aa599a0c1283ad

File: client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+			slug: PropTypes.string,
+			ID: PropTypes.number,
+		} ),
+		className: PropTypes.string,
+	};
+
+	componentDidMount = () => {
+		const { site } = this.props;
+
+		if ( site && site.ID ) {
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js]: Agreed! Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/054bd7f282edba71c5f26c8377aa599a0c1283ad

File: client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+			userBeganEditing: false,
+		};
+	}
+
+	static propTypes = {
+		site: PropTypes.shape( {
+			slug: PropTypes.string,
+			ID: PropTypes.number,
+		} ),
+		className: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-wcs.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/8659592406fea67b570072721ac972d78f64c586

File: client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+		} ),
+	};
+
+	onDeactivate = event => {
+		event.preventDefault();
+	};
+
+	render = () => {
+		const { className, site, translate } = this.props;
+
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js]: It should not be null, but defensively changed site to isRequired in https://github.com/Automattic/wp-calypso/pull/19535/commits/8659592406fea67b570072721ac972d78f64c586

File: client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	static propTypes = {
+		className: PropTypes.string,
+		site: PropTypes.shape( {
+			options: PropTypes.shape( {
+				admin_url: PropTypes.string.isRequired,
+			} ),
+		} ),
+	};
+
+	onDeactivate = event => {
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js]: Oops - fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/8659592406fea67b570072721ac972d78f64c586

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 8]
```diff
@@ -9,207 +9,90 @@ import PropTypes from 'prop-types';
...
-
 	static propTypes = {
-		site: PropTypes.shape( {
-			slug: PropTypes.string,
-			ID: PropTypes.number,
-		} ),
 		className: PropTypes.string,
+		siteId: PropTypes.number,
+		sitePluginsLoaded: PropTypes.bool,
+		taxJarPluginActive: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/index.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/8659592406fea67b570072721ac972d78f64c586

**Comment by allendav** [comment]
@timmyc wrote:

> I uncheck the "charge taxes on shipping costs", save, am told things are all good, but upon refresh the old setting is still shown. Note, this is also happening in production

I can't reproduce this with this branch - are you sure the TaxJar plugin is not still active on your site? (It forces these settings' values and can cause this behavior, which is basically why we are doing this PR.)

**Comment by jameskoster** [comment]
> Sales tax calculations are provided by WooCommerce Services. When this option is enabled, WooCommerce Services will share some of your data with a third party. Learn more.

I thought we were averse to highlighting 'WooCommerce' anywhere in Store? 

I actually think it's the other card which is the problem. "Your store's taxes are being handled by a third-party plugin" just feels too vague. It reminds me of one of those unhelpful Windows error messages from years gone by xD

Couldn't we be more explicit there? Explain that they're running the TaxJar plugin which is no longer required unless they want to do x & y. Then re-assure them that if they deactivate it their store will continue to function just fine.

**Comment by timmyc** [comment]
@allendav still seeing the issue on my site with this branch - another gif from this morning:

tax-settings-branch

TaxJar installed but disabled:

plugins_ _timmy_s_flies_ _wordpress

**Comment by allendav** [comment]
@jameskoster wrote:

> I thought we were averse to highlighting 'WooCommerce' anywhere in Store?

I think we need too not be afraid to mention WCS - after all, WooCommerce Services is a product we want to sell through Store on WPCOM. cc @toddwilkens

**Comment by allendav** [comment]
@timmyc wrote:

> still seeing the issue on my site with this branch - another gif from this morning

@timmyc - a few questions (or grant me access to that site)

1. Are you able to toggle the setting (and get it to stick after save) if tax calculations are enabled?
2. In wp-admin on that site ( /wp-admin/admin.php?page=wc-settings&tab=tax ), when tax calculations are enabled, are you able to change the Shipping Tax Class setting from Zero Rate to Standard and have it stick? What appears in calypso after a refresh?
3. In wp-admin on that site, when tax calculations are enabled, are you able to change the Shipping Tax Class setting from Standard to Zero Rate and have it stick? What appears in calypso after a refresh?

**Comment by allendav** [comment]
@jameskoster wrote:

> Couldn't we be more explicit there? Explain that they're running the TaxJar plugin which is no longer required unless they want to do x & y. Then re-assure them that if they deactivate it their store will continue to function just fine.

Good point. OK, let's do this then. If the TaxJar plugin is active on their site, we'll show this

screen shot 2017-11-14 at 10 02 17 am

Sound good?

File: client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+									},
+								}
+							) }
+						</p>
+						<p>
+							{ translate(
+								'You can manage settings, including your personal TaxJar API key, in the plugin settings. '
+							) }
+							<ExternalLink icon href={ adminUrl } rel="noopener noreferrer">
+								{ translate( 'Plugin settings' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js]: Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Edit plugin settings' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by timmyc** [comment]
@allendav I just sent you an invite to my test site. Feel free to do whatever you please there... make yourself at home 🏚

**Comment by jameskoster** [comment]
@allendav I think this is an improvement. It's a bit wordy but it's all relevant. 

Perhaps to break things up a little we should bullet-list the features the plugin adds. Then switch the last 2 paragraphs.

> Your store’s taxes are being managed by the TaxJar - Sales Tax Automation for WooCommerce plugin. This plugin provides optional features such as:

> * Automated reporting & filing
> * Multi-nexus support

> If you don’t need these features we recommend removing this plugin. Don’t worry, we’ll still calculate sales taxes for you.

> You can manage this plugin’s settings in wp-admin.

**Comment by allendav** [comment]
@jameskoster - with https://github.com/Automattic/wp-calypso/pull/19535/commits/806731ba217c2754e2810d4b35a5eeef65ea1fe5 we now have

screen shot 2017-11-15 at 11 24 09 am

File: client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+										a: <a href={ pluginUrl } className="taxes__tax-jar-deactivate" />,
+										b: <strong />,
+									},
+								}
+							) }
+						</p>
+						<p>
+							{ translate( "You can manage this plugin's settings in wp-admin. " ) }
+							<ExternalLink icon href={ adminUrl } rel="noopener noreferrer">
+								{ translate( 'Plugin settings' ) }
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/settings/taxes/taxes-taxjar.js]: Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Edit plugin settings' )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by allendav** [comment]
@timmyc - I tracked down your lack-of-stickyness issue and wrote it up separately here - https://github.com/Automattic/wp-calypso/issues/19857 - not sure how your site got in that state

**Comment by jameskoster** [comment]
@allendav one final nitpick; let's lose the bottom margin on the final paragraph :)

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 11]
```diff
@@ -9,207 +9,92 @@ import PropTypes from 'prop-types';
...
+		isRequestingSitePlugins: PropTypes.bool,
+		siteId: PropTypes.number,
+		sitePluginsLoaded: PropTypes.bool,
+		siteSlug: PropTypes.string,
+		taxJarPluginActive: PropTypes.bool,
 	};
 
 	componentDidMount = () => {
-		const { site } = this.props;
+		const { isRequestingSitePlugins, siteId, sitePluginsLoaded } = this.props;
```
Comment (timmyc) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: This logic is duplicated in `componentWillReceiveProps`. Perhaps it can be extracted into a `maybeFetchPlugins` method and called in both lifecycle methods?

File: client/extensions/woocommerce/app/settings/taxes/index.js [hunk 12]
```diff
@@ -9,207 +9,92 @@ import PropTypes from 'prop-types';
...
 
-		if ( site && site.ID ) {
-			this.props.fetchTaxSettings( site.ID );
+		if ( siteId && ! isRequestingSitePlugins && ! sitePluginsLoaded ) {
```
Comment (timmyc) [inline client/extensions/woocommerce/app/settings/taxes/index.js]: The only thing that I thought could go awry here is that potentially site plugins could become stale in the context of this component. Since `sitePluginsLoaded` just checks to see if the plugins array is empty, we could potentially not fetch the latest plugin data here... i.e. I change plugin status outside of calypso.

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js [hunk 13]
```diff
@@ -41,25 +41,21 @@ import QuerySettingsGeneral from 'woocommerce/components/query-settings-general'
...
+		if ( loadedSettingsGeneral && ! loadedTaxRates ) {
+			this.props.fetchTaxRates( siteId, address );
 		}
 	};
 
 	componentWillReceiveProps = nextProps => {
 		if ( nextProps.loadedSettingsGeneral ) {
 			if ( ! nextProps.loadedTaxRates ) {
-				this.props.fetchTaxRates( nextProps.site.ID, nextProps.address );
+				this.props.fetchTaxRates( nextProps.siteId, nextProps.address );
```
Comment (timmyc) [inline client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: could these two `if` conditions be combined here? `if ( nextProps.loadedSettingsGeneral && ! nextProps.loadedTaxRates )`

**Comment by allendav** [comment]
@jameskoster wrote:

> @allendav one final nitpick; let's lose the bottom margin on the final paragraph :)

Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/8d4fab173e45750ad6615570e47d775225610cd9

screen shot 2017-11-16 at 9 06 17 am

File: client/extensions/woocommerce/app/settings/taxes/taxes-rates.js (hunk 13 above)
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/taxes-rates.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/4dc13074b82efbf815f60cf9475d4ecdb8eaf096

File: client/extensions/woocommerce/app/settings/taxes/index.js (hunk 11 above)
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/index.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/19535/commits/4dc13074b82efbf815f60cf9475d4ecdb8eaf096

File: client/extensions/woocommerce/app/settings/taxes/index.js (hunk 12 above)
Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/index.js]: If you are on the Settings > Taxes page, then go to Plugins and install and activate TaxJar, then return to the Settings > Taxes page, you'll have this problem too. I'll come up with a fix.

Comment (timmyc) [reply client/extensions/woocommerce/app/settings/taxes/index.js]: It isn't super critical imho, but figured it is worth noting. Could be handled in a subsequent PR.

Comment (allendav) [reply client/extensions/woocommerce/app/settings/taxes/index.js]: OK, in https://github.com/Automattic/wp-calypso/pull/19535/commits/2e6ebd33b7e8c239042ba2e773448cabf754c007 I made a change so that if you install/activate TaxJar from Calypso and return to the Taxes Settings view, at least we'll update appropriately.  It doesn't solve the problem of navigating to wp-admin and messing there, but it is a start.

I've written https://github.com/Automattic/wp-calypso/issues/19928 for the case where they use another window/tab to deactivate/activate the TaxJar plugin.

**Comment by allendav** [comment]
@timmyc - still good to go with the last round of changes?

**Comment by timmyc** [comment]
Yep @allendav - still looking good 👍

**Comment by yansern** [comment]
Please ignore my previous mention as it was incorrectly referenced.
